feat(keywordprg): use :terminal for external commands

gpanders opened the pull request:
Closes #2995.

closes https://github.com/neovim/neovim/issues/2761

gpanders committed:
feat: use :terminal for external commands in &keywordprg

External commands are opened in a terminal buffer in a new tab. <Esc> is
mapped to stop the job and close the buffer.

Closes #2995.

justinmk commented:
python files are a way to test this qualitatively: https://github.com/neovim/neovim/blob/7215d356949bf07960cc3a3d8b83772635d729d0/runtime/ftplugin/python.vim#L127-L131

Technically we need to go a bit further to mimic the old `:!` behavior, e.g.
- enter terminal-mode ("insert" mode) automatically
- add a buffer-local `:tnoremap <esc>` which kills the buffer
- restore the original buffer and its `@#` (alternative) buffer

gpanders commented:
> 1 test failing:
> 
> ```
> test/functional/ui/messages_spec.lua:1225: fails on unix CI
> [  ERROR   ] 1 error, listed below:
> [  ERROR   ] test/functional/normal/K_spec.lua @ 27: K invokes non-prefixed 'keywordprg' as shell command
> test/functional/helpers.lua:111: Vim:E484: Can't open file K_spec_out
> ```

I'm looking into this. It looks like this is happening because `:term` is asynchronous while `:!` is not, e.g. the test for equality in `K_spec.lua` is happening before the `:term {cmd}` finishes executing. I guess we need a way to signal to the test to wait until the terminal command finishes, but I'm not yet sure how to accomplish that.

justinmk commented:
> because `:term` is asynchronous while `:!` is not, e.g. the test for equality in `K_spec.lua` is happening before the `:term {cmd}` finishes executing. I guess we need a way to signal to the test to wait until the terminal command finishes

can use retry() https://github.com/neovim/neovim/blob/7215d356949bf07960cc3a3d8b83772635d729d0/test/functional/autocmd/termxx_spec.lua#L23

gpanders commented:
> Technically we need to go a bit further to mimic the old `:!` behavior, e.g.
> 
>     * enter terminal-mode ("insert" mode) automatically
> 
>     * add a buffer-local `:tnoremap <esc>` which kills the buffer
>       
>       * restore the original buffer and its `@#` (alternative) buffer
>       * or open a new tab and close it

I like the new tab approach, so I went with that. I tested this with pydoc as you suggested and it works pretty well for me.

gpanders committed:
test: update K_spec test

:term runs asynchronously so we need to wait to run the equality test
until the job finishes.

gpanders commented:
There is still a failing test in `test_normal.vim`:

https://github.com/neovim/neovim/blob/master/src/nvim/testdir/test_normal.vim#L1243-L1250

This is failing because `execute('term ...')` doesn't return a string like `execute('!...')` does.

The `K_spec.lua` test already checks for this case, so can we just delete this one?

justinmk commented on src/nvim/normal.c:
```diff
@@ -5034,6 +5035,15 @@ static void nv_ident(cmdarg_T *cap)
+
+      add_map((char_u *)"<buffer> <esc> <Cmd>call jobstop(&channel)<CR>", TERM_FOCUS, true);
+      do_autocmd_event(EVENT_TERMCLOSE, (char_u *)"<buffer>", false, false, (char_u *)"bdelete!",
+                       false, AUGROUP_DEFAULT);
```
can we just use `do_cmdline_cmd` here? That avoids exposing do_autocmd_event(). Performance of `do_cmdline_cmd` is small compared to the cost of the rest of this codepath.

gpanders replied:
Sure, I'll just use `do_autocmd` (it parses a string like `do_cmdline_cmd` so is equally readable).

justinmk commented on src/nvim/testdir/test_normal.vim:
```diff
@@ -1226,29 +1226,6 @@ func Test_normal23_K()
   set iskeyword-=%
   set iskeyword-=\|
 
-  " Only expect "man" to work on Unix
```
Avoid deleting lines in the oldtests. It makes merging upstream changes much easier if the lines exist, but are either commented-out (for 1-2 lines) or wrapped in a guard (many lines, such as this case).

To avoid indenting the skipped lines (again to avoid merge-conflicts from upstream), we can do an early-return here:

```
if has('nvim')  " Nvim K uses :terminal. #15398
  let &keywordprg = k
  bw!
endif
```

justinmk commented on src/nvim/testdir/test_normal.vim:
```diff
@@ -1226,29 +1226,6 @@ func Test_normal23_K()
   set iskeyword-=\|
 
-  " Only expect "man" to work on Unix
-  if !has("unix")
```
Avoid deleting lines in the oldtests. It makes merging upstream changes much easier if the lines exist, but are either commented-out (for 1-2 lines) or skipped.

To avoid indenting the skipped lines (again to avoid merge-conflicts from upstream), we can do an early-return here:

```
if !has("unix") || has('nvim')  " Nvim K uses :terminal. #15398
```

gpanders committed:
test: skip external keywordprg test in test_normal

This is already covered by K_spec.lua.

justinmk commented:
This is working pretty nicely, but behavior is weird if `K` is performed on an unknown symbol. Try it in a python file on a random symbol that pydoc doesn't understand. It just closes without any message. Can we either (1) prevent it from closing (show whatever pydoc exit code/message), or (2) check the exit code and show a generic message?

I guess we need https://github.com/neovim/neovim/pull/15406 first?

gpanders commented:
> This is working pretty nicely, but behavior is weird if `K` is performed on an unknown symbol. Try it in a python file on a random symbol that pydoc doesn't understand. It just closes without any message. Can we either (1) prevent it from closing (show whatever pydoc exit code/message), or (2) check the exit code and show a generic message?
> 
> I guess we need #15406 first?

Yep, #15406 dovetails with this one quite nicely. We can add the v:event.status check to the TermClose autcommand.

gpanders committed:
feat: only close keywordprg buffer if command exited without errors

gpanders commented:
> > This is working pretty nicely, but behavior is weird if `K` is performed on an unknown symbol. Try it in a python file on a random symbol that pydoc doesn't understand. It just closes without any message. Can we either (1) prevent it from closing (show whatever pydoc exit code/message), or (2) check the exit code and show a generic message?
> > I guess we need #15406 first?
> 
> Yep, #15406 dovetails with this one quite nicely. We can add the v:event.status check to the TermClose autcommand.

Hrm, so at least with `pydoc` it looks like it still returns with exit code 0 even on an unknown symbol, it just returns early. However I consider this a problem with `pydoc` specifically (they really should be returning a non-zero exit code in such cases, I think) so I think we should still move forward.

justinmk commented on src/nvim/normal.c:
```diff
@@ -4931,7 +4931,8 @@ static void nv_ident(cmdarg_T *cap)
 
-      STRCAT(buf, "! ");
+      do_cmdline_cmd("tabnew");
+      STRCAT(buf, "term ");
```
avoid abbreviated ex command names in scripts/core. Does the calcuatlion of `buf_size` need to change?

gpanders replied:
>  Does the calcuatlion of buf_size need to change?

It's currently `n*2 + 30 + STRLEN(&keywordprg)` where `n` is the length of the identifier being queried. I don't know how or why this number was decided on for the buffer size but we are still within these bounds.

justinmk commented on src/nvim/normal.c:
```diff
@@ -5028,6 +5029,14 @@ static void nv_ident(cmdarg_T *cap)
+      restart_edit = 'i';
+
+      add_map((char_u *)"<buffer> <esc> <Cmd>call jobstop(&channel)<CR>", TERM_FOCUS, true);
+      do_cmdline_cmd("autocmd TermClose <buffer> bdelete!");
```
even if it doesn't apply to pydoc, doesn't it make sense to skip this for non-zero `v:event.status` ?

gpanders replied:
Yes, I have that change locally and was testing it when I made the discovery about pydoc. I'll push it soon.

justinmk commented on test/functional/normal/K_spec.lua:
```diff
@@ -31,7 +31,7 @@ describe('K', function()
     feed('i'..test_file..'<ESC>K')
-    feed('<CR>') -- Press ENTER
+    retry(nil, nil, function() eq(1, eval('filereadable("'..test_file..'")')) end)
     eq({'fnord'}, eval("readfile('"..test_file.."')"))
```
Just for reference: it's also fine to put retry() around the test assertion. It'll fail correctly/informatively.

justinmk commented on test/functional/normal/K_spec.lua:
```diff
@@ -1,6 +1,6 @@
 local helpers = require('test.functional.helpers')(after_each)
-local eq, clear, eval, feed =
-  helpers.eq, helpers.clear, helpers.eval, helpers.feed
+local eq, neq, clear, eval, feed, retry =
+  helpers.eq, helpers.neq, helpers.clear, helpers.eval, helpers.feed, helpers.retry
```
```suggestion
local eq, clear, eval, feed, retry =
  helpers.eq, helpers.clear, helpers.eval, helpers.feed, helpers.retry
```

justinmk committed:
Update test/functional/normal/K_spec.lua
